### ISSUE TITLE
Copying signature from an overloaded function results in error

### ISSUE THREAD

**Issue by InSyncWithFoo** [microsoft/pyright#7221]
**Reproducible minimal example**

I have this decorator factory which copies a function's signature to another ([playground link](https://pyright-play.net/?strict=true&code=GYJw9gtgBAxmA28CmMAuBLMA7AzgOgEMAjGKdCABzBFSgGEDFjkBYAKFEilQE8L0sAczKVqtAIJYeAGlgEcqWQAUCIAhADKFFLIAqfJADVV7drqgBeKPu3GQACgDkuxwEp2Sy1BVrN2mE5KbqZsACZIwLBgFDwA%2BjjoglgEqACuIEixnBD2sdSJAgBc9IzwzEgA2kp6ALquUAC0AHwlTETIFRUMbR14fbKSPDU1st1l7ZXV1sOF7FDzUOGR4XBqqNT2qKqCSKjFY%2BUVfXgDUnWNLQcTVbWzbAsPUBlpIFhyCvZXHVO6I9zbu3c9wWcwWz3SbxW1BS1BC2W4fAEwnIVBoUDAADckCB4GACKFZAAZdCobGMEIAAUx2Nx%2BPYSyggnsGOKxNJangFQAjOdmmQsHtQfNjkKoOwqVicXjQvSIozmaySWTOQAmXktBQgO4PEVsWWRJksqBs5Xc2Rq%2Bp8gS0AA%2BUE12oWuvFcBi8USyRemWy9kEQIZgliAHc1BRtA4AFTbHDFQayCMRgDWQejsaklpagzuAGIFgA-AuFwuWFrY8BaqAAIkDIYIYexlbIOCgEFUiaQoSg8nRktpBKgRFStCwYBEFGQECQApSmDe6GbFHAGPQ4RlwOFfRCGSxjFivG0vvsXNc9UeuYgfB4UAAZFAYiBEgALWgWFrW9jbpC7-dIQ8qk%2BPFA56Xjed48A%2BgjPiW9qoCA7BAA)):

```py
from collections.abc import Callable
from typing import Any, cast, ParamSpec, TypeVar

T = TypeVar('T')
P = ParamSpec('P')

def copy_signature_from(_origin: Callable[P, T]) -> Callable[[Callable[..., Any]], Callable[P, T]]:
    def decorator(target: Callable[..., Any]) -> Callable[P, T]:
        return cast(Callable[P, T], target)
    
    return decorator
```

It works in normal cases. However, if the origin function is `@overload`-ed, Pyright emits an error ([mypy doesn't](https://mypy-play.net/?mypy=latest&python=3.12&gist=6739902ee494cae71f4c2fd72927625d&flags=strict)):

```py
from typing import overload, Literal

@overload
def g(v: Literal[1]) -> int:
    ...
    
@overload
def g(v: Literal[2]) -> str:
    ...

def g(v: Literal[1, 2]) -> int | str:
    ...

@copy_signature_from(g)
def g_wrapper(*args: Any, **kwargs: Any) -> Any:
#   ~~~~~~~~~ => error: "g_wrapper" is marked as overload, but no implementation is provided
    ...
```

Regardless, the return type <em>is</em> deduced correctly:

```py
reveal_type(g(1))      # mypy & pyright => int
reveal_type(g(2))      # mypy & pyright => str
```

**Comment by erictraut** [comment]
Thanks for the bug report. This will be fixed in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.351, which I just published.
